src/core/or/or.h

```diff
@@ -841,12 +845,9 @@ typedef struct protover_summary_flags_t {
   unsigned int supports_v3_rendezvous_point: 1;
 
   /** True iff this router has a protocol list that allows clients to
-   * negotiate hs circuit setup padding. Requires Padding>=2. */
+   * negotiate hs circuit setup padding. Requires Padding=2. */
```

Comment:
Good catch that our current behavior is =, not >=.  But do you suppose that our behavior should be >= instead?  If so I can open a ticket to write a proposal.

Reply:
There are a few instances of ">=" and "=" confusion in protocol versions. We should try to fix them all.

It only matters when we remove protocol versions. We haven't really specified, tested, or exercised this functionality in practice. And so our reviewers lack experience. (And when we did discover a need for it with NSS and LinkAuth, it was more complicated than we expected.)

I'd like to see a proposal that tells us how to check future protocol versions as they are added. Along with a migration plan for disabling dependent and orthogonal protocol versions.

I originally thought we might want an "all future versions" check. But that's a bad idea, because that check would make it impossible to add a version that disables a feature. (In that case, we would choose a new name for the new feature.)

So let's also open a ticket to use the versions that currently support the feature being tested. We should:
* replace all ">=" checks (unless there is a specific exception), and
* replace all "=" checks.
Let's make sure we check all the places where we use protocol versions, even if they don't have a summary flag.

There's only one ">=" check exception right now: ed25519 link specifiers for extends. We could assume that any future link authentication method would use ed25519 keys, or safely ignore them. Or we could assume that any future protocol that dropped ed25519 keys for direct connections, would also make them unsuitable for extends.

Let's try to avoid designs where we have to predict the future :-)

Overall, I think it would be helpful if future protocol versions were orthogonal. Or if they depend on earlier features, that dependency should be clearly documented. (For example, Relay=3 IPv6 extends depends on Relay=2 EXTEND2 cells. So if we were checking EXTEND2 cell support, it would be Relay=2 or Relay=3.)

Reply:
I've changed my mind, I think we want to check for "any later protocol version".

See my tor-dev email here:
https://lists.torproject.org/pipermail/tor-dev/2020-May/014297.html